File: src/okbay/theme.py
"""Resolve Omarchy system theme colors for Atlas and QML surfaces."""
from __future__ import annotations
import os, time
from pathlib import Path

SWITCHBAY_TYPE = {
    "project": "#4d1ae8", "projects": "#4d1ae8",
    "analysis": "#1d6996", "analyses": "#1d6996",
    "concept": "#38a6a5", "concepts": "#38a6a5",
    "entity": "#0f8554", "entities": "#0f8554",
    "evidence": "#73af48",
    "fact": "#edad08", "facts": "#edad08",
    "figure": "#e17c05", "figures": "#e17c05",
    "table": "#cc503e", "tables": "#cc503e",
    "source": "#94346e", "sources": "#94346e",
    "note": "#6f4070", "notes": "#6f4070",
    "todo": "#9656a2", "todo-list": "#9656a2",
    "hub": "#6be8b3", "unclassified": "#ffffff",
    "missing": "#5a6068", "default": "#9aa0a8",
}
SWITCHBAY_CHROME = {
    "mode": "dark", "accent": "#6be8b3", "selection": "#1f242c", "muted": "#343942",
    "background": "#0f1115", "dark_background": "#0b0d10", "darker_background": "#07080a",
    "lighter_background": "#161a21", "foreground": "#e6e8eb", "dark_foreground": "#5a6068",
    "light_foreground": "#c5c9ce", "bright_foreground": "#f4f6f8",
    "red": "#cc503e", "yellow": "#edad08", "orange": "#e17c05", "green": "#0f8554",
    "cyan": "#38a6a5", "blue": "#1d6996", "magenta": "#94346e", "brown": "#6f4070",
    "bright_red": "#e17c05", "bright_yellow": "#edad08", "bright_green": "#73af48",
    "bright_cyan": "#6be8b3", "bright_blue": "#4d1ae8", "bright_magenta": "#9656a2",
}
KIND_TO_NAMED = {
    "project": "bright_blue", "projects": "bright_blue",
    "analysis": "blue", "analyses": "blue",
    "concept": "cyan", "concepts": "cyan",
    "entity": "green", "entities": "green",
    "evidence": "bright_green", "fact": "yellow", "facts": "yellow",
    "figure": "orange", "figures": "orange",
    "table": "red", "tables": "red",
    "source": "magenta", "sources": "magenta",
    "note": "brown", "notes": "brown",
    "todo": "bright_magenta", "todo-list": "bright_magenta",
    "hub": "accent", "unclassified": "bright_foreground",
    "missing": "muted", "default": "dark_foreground",
}

def _home() -> Path:
    return Path(os.environ.get("HOME") or Path.home())

def omarchy_state_root() -> Path:
    raw = os.environ.get("XDG_STATE_HOME")
    state = Path(raw) if raw else _home() / ".local" / "state"
    return state / "omarchy" / "current"

def theme_name() -> str:
    for candidate in (omarchy_state_root() / "theme.name", _home() / ".config" / "omarchy" / "current" / "theme.name"):
        if candidate.is_file():
            name = candidate.read_text(encoding="utf-8").strip()
            if name:
                return name
    return ""

def colors_toml_path() -> Path | None:
    for p in (
        omarchy_state_root() / "theme" / "colors.toml",
        omarchy_state_root() / "next-theme" / "colors.toml",
        _home() / ".config" / "omarchy" / "current" / "theme" / "colors.toml",
        _home() / ".config" / "omarchy" / "themes" / "switchbay" / "colors.toml",
    ):
        if p.is_file():
            return p
    return None

def parse_toml_colors(text: str) -> dict[str, str]:
    out: dict[str, str] = {}
    for raw in text.splitlines():
        line = raw.split("#", 1)[0].strip()
        if not line or "=" not in line:
            continue
        key, val = line.split("=", 1)
        key, val = key.strip(), val.strip().strip('"').strip("'")
        if key and val:
            out[key] = val
    return out

def load_omarchy_colors() -> dict[str, str]:
    path = colors_toml_path()
    merged = dict(SWITCHBAY_CHROME)
    if path is None:
        return merged
    try:
        merged.update(parse_toml_colors(path.read_text(encoding="utf-8")))
    except OSError:
        pass
    return merged

def _kind_color(colors: dict[str, str], kind: str) -> str:
    kind = (kind or "default").lower().replace(" ", "-")
    explicit = colors.get(f"type_{kind}") or colors.get(f"type-{kind}")
    if explicit:
        return explicit
    named = KIND_TO_NAMED.get(kind)
    if named and colors.get(named):
        return colors[named]
    return SWITCHBAY_TYPE.get(kind) or colors.get("dark_foreground") or SWITCHBAY_TYPE["default"]

def type_palette(colors=None):
    colors = colors or load_omarchy_colors()
    kinds = set(SWITCHBAY_TYPE) | set(KIND_TO_NAMED)
    return {kind: _kind_color(colors, kind) for kind in kinds}

def atlas_theme() -> dict:
    colors = load_omarchy_colors()
    types = type_palette(colors)
    path = colors_toml_path()
    return {
        "name": theme_name() or "switchbay-fallback",
        "source": str(path) if path else "",
        "mode": colors.get("mode", "dark"),
        "chrome": {k: colors.get(k, SWITCHBAY_CHROME.get(k, "")) for k in SWITCHBAY_CHROME if k != "mode"},
        "types": types,
        "ts": time.time(),
    }

def resolve() -> dict:
    theme = atlas_theme()
    theme["kinds"] = theme.get("types") or {}
    theme["theme"] = theme.get("name") or "switchbay-fallback"
    chrome = theme.get("chrome") or {}
    theme["accent"] = chrome.get("accent", "#6be8b3")
    theme["background"] = chrome.get("background", "#0f1115")
    theme["foreground"] = chrome.get("foreground", "#e6e8eb")
    return theme

def kind_colors(colors=None):
    return type_palette(colors)

File: crates/okbayd/src/theme.rs
use serde_json::{json, Map, Value};
use std::fs;
use std::path::PathBuf;

fn home() -> PathBuf {
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("/tmp"))
}

fn colors_toml_path() -> Option<PathBuf> {
    let state = std::env::var_os("XDG_STATE_HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| home().join(".local/state"));
    let cands = [
        state.join("omarchy/current/theme/colors.toml"),
        home().join(".config/omarchy/current/theme/colors.toml"),
        home().join(".config/omarchy/themes/switchbay/colors.toml"),
    ];
    cands.into_iter().find(|p| p.is_file())
}

fn parse_toml_colors(text: &str) -> std::collections::BTreeMap<String, String> {
    let mut out = std::collections::BTreeMap::new();
    for raw in text.lines() {
        let line = raw.split('#').next().unwrap_or("").trim();
        if let Some((k, v)) = line.split_once('=') {
            let val = v.trim().trim_matches('"').trim_matches('\'').to_string();
            // Empty values must not blank Switchbay fallbacks (CE Atlas palette).
            if !val.is_empty() {
                out.insert(k.trim().to_string(), val);
            }
        }
    }
    out
}

fn color(
    colors: &std::collections::BTreeMap<String, String>,
    keys: &[&str],
    fallback: &str,
) -> String {
    for k in keys {
        if let Some(v) = colors.get(*k) {
            if !v.is_empty() {
                return v.clone();
            }
        }
    }
    fallback.to_string()
}

/// Full Switchbay CE type palette (parity with Python theme.SWITCHBAY_TYPE).
fn type_palette(colors: &std::collections::BTreeMap<String, String>) -> Map<String, Value> {
    let pairs: &[(&str, &[&str], &str)] = &[
        ("project", &["type_project", "type-project", "bright_blue"], "#4d1ae8"),
        ("projects", &["type_project", "type-project", "bright_blue"], "#4d1ae8"),
        ("analysis", &["type_analysis", "type-analysis", "blue"], "#1d6996"),
        ("analyses", &["type_analysis", "type-analysis", "blue"], "#1d6996"),
        ("concept", &["type_concept", "type-concept", "cyan"], "#38a6a5"),
        ("concepts", &["type_concept", "type-concept", "cyan"], "#38a6a5"),
        ("entity", &["type_entity", "type-entity", "green"], "#0f8554"),
        ("entities", &["type_entity", "type-entity", "green"], "#0f8554"),
        ("evidence", &["type_evidence", "type-evidence", "bright_green"], "#73af48"),
        ("fact", &["type_fact", "type-fact", "yellow"], "#edad08"),
        ("facts", &["type_fact", "type-fact", "yellow"], "#edad08"),
        ("figure", &["type_figure", "type-figure", "orange"], "#e17c05"),
        ("figures", &["type_figure", "type-figure", "orange"], "#e17c05"),
        ("table", &["type_table", "type-table", "red"], "#cc503e"),
        ("tables", &["type_table", "type-table", "red"], "#cc503e"),
        ("source", &["type_source", "type-source", "magenta"], "#94346e"),
        ("sources", &["type_source", "type-source", "magenta"], "#94346e"),
        ("note", &["type_note", "type-note", "brown"], "#6f4070"),
        ("notes", &["type_note", "type-note", "brown"], "#6f4070"),
        ("todo", &["type_todo", "type-todo", "bright_magenta"], "#9656a2"),
        ("todo-list", &["type_todo", "type-todo-list", "bright_magenta"], "#9656a2"),
        ("hub", &["type_hub", "type-hub", "accent"], "#6be8b3"),
        ("unclassified", &["type_unclassified", "bright_foreground"], "#ffffff"),
        ("missing", &["type_missing", "muted"], "#5a6068"),
        ("default", &["dark_foreground"], "#9aa0a8"),
    ];
    let mut kinds = Map::new();
    for (name, keys, fb) in pairs {
        kinds.insert((*name).to_string(), json!(color(colors, keys, fb)));
    }
    kinds
}

pub fn atlas_theme() -> Value {
    let mut colors = std::collections::BTreeMap::from([
        ("accent".into(), "#6be8b3".into()),
        ("background".into(), "#0f1115".into()),
        ("foreground".into(), "#e6e8eb".into()),
        ("muted".into(), "#343942".into()),
        ("dark_foreground".into(), "#5a6068".into()),
        ("light_foreground".into(), "#9aa0a8".into()),
        ("bright_foreground".into(), "#f4f6f8".into()),
        ("selection".into(), "#1f242c".into()),
        ("dark_background".into(), "#0c0e12".into()),
        ("lighter_background".into(), "#161a21".into()),
        ("darker_background".into(), "#08090c".into()),
        ("red".into(), "#cc503e".into()),
        ("yellow".into(), "#edad08".into()),
        ("orange".into(), "#e17c05".into()),
        ("green".into(), "#0f8554".into()),
        ("cyan".into(), "#38a6a5".into()),
        ("blue".into(), "#1d6996".into()),
        ("magenta".into(), "#94346e".into()),
        ("brown".into(), "#6f4070".into()),
        ("bright_red".into(), "#e17c05".into()),
        ("bright_yellow".into(), "#edad08".into()),
        ("bright_green".into(), "#73af48".into()),
        ("bright_cyan".into(), "#6be8b3".into()),
        ("bright_blue".into(), "#4d1ae8".into()),
        ("bright_magenta".into(), "#9656a2".into()),
    ]);
    let source = colors_toml_path();
    if let Some(p) = &source {
        if let Ok(text) = fs::read_to_string(p) {
            colors.extend(parse_toml_colors(&text));
        }
    }
    let chrome = json!({
        "accent": color(&colors, &["accent"], "#6be8b3"),
        "background": color(&colors, &["background"], "#0f1115"),
        "foreground": color(&colors, &["foreground"], "#e6e8eb"),
        "muted": color(&colors, &["muted"], "#343942"),
        "dark_foreground": color(&colors, &["dark_foreground"], "#5a6068"),
        "light_foreground": color(&colors, &["light_foreground"], "#9aa0a8"),
        "selection": color(&colors, &["selection"], "#1f242c"),
        "dark_background": color(&colors, &["dark_background"], "#0c0e12"),
        "lighter_background": color(&colors, &["lighter_background"], "#161a21"),
        "darker_background": color(&colors, &["darker_background"], "#08090c"),
    });
    let kinds = Value::Object(type_palette(&colors));
    let mut css = serde_json::Map::new();
    if let Some(obj) = chrome.as_object() {
        for (k, v) in obj {
            if let Some(s) = v.as_str() {
                css.insert(format!("--{}", k.replace('_', "-")), json!(s));
            }
        }
    }
    css.insert("--accent".into(), chrome["accent"].clone());
    json!({
        "name": "omarchy-or-switchbay",
        "source": source.map(|p| p.display().to_string()).unwrap_or_default(),
        "chrome": chrome,
        "kinds": kinds,
        "types": kinds,
        "css": css,
        "theme": "omarchy",
        "mode": colors.get("mode").cloned().unwrap_or_else(|| "dark".into()),
    })
}
